actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -904,14 +904,34 @@ def permitted_scalar?(value)
         PERMITTED_SCALAR_TYPES.any? { |type| value.is_a?(type) }
       end
 
-      def permitted_scalar_filter(params, key)
-        if has_key?(key) && permitted_scalar?(self[key])
-          params[key] = self[key]
+      # Adds existing keys to the params if their values are scalar.
+      #
+      # For example:
+      #
+      #   puts self.keys #=> ["zipcode(90210i)"]
+      #   params = {}
+      #
+      #   permitted_scalar_filter(params, "zipcode")
+      #
+      #   puts params.keys # => ["zipcode"]
+      def permitted_scalar_filter(params, permitted_key)
+        if has_key?(permitted_key) && permitted_scalar?(self[permitted_key])
+          params[permitted_key] = self[permitted_key]
         end
 
-        keys.grep(/\A#{Regexp.escape(key)}\(\d+[if]?\)\z/) do |k|
```

Comment:
If the important part is to avoid the dynamic regex for "simple" params where it never actually matches, how about:

```ruby
keys.grep(/\(\d+[if]?\)\z/) do |key|
  next unless $~.pre_match == permitted_key

  params[key] = self[key] if permitted_scalar?(self[key])
end
```

It'll allocate more when there _is_ a match, which your `match?` formulation would not, but feels like it might be a better performance vs legibility trade-off?

Reply:
`each`+`if`, then? The deferred assignment just seems so awkwardly back-to-front.

Reply:
This code is way harder to read. If I'd have to work on this method I'd fear to change it. And I'm saying this as a core team member, imagine a first-time contributor. I prefer Matthew's suggestion even if that means a slower method.

Reply:
No one has touched this method in 6 years. It will be fine.

Reply:
Regarding the slower case on multiple matches, in my case when I see one scalar match i see multiple for example `per_page` and `page`.

Reply:
No, sorry it will not be fine. I really thing code readability is better than speed in an open source project. Of course I'd prefer to have speed, but not in detriment to code maintainability.

Reply:
closing this pr then

Reply:
Wait, doesn't we have performance improments by using the following version?

```ruby
keys.grep(/\(\d+[if]?\)\z/) do |key|
  next unless $~.pre_match == permitted_key

  params[key] = self[key] if permitted_scalar?(self[key])
end
```

Reply:
Just to be clear I don't think your change were bad, just that we could have a middle ground here. Matthew's suggestion is still readable but faster (I believe, since I didn't measured) than our current implementation. Of course still allocates an array, but I'm not sure we should sacrifice readability to avoid this array.